### ISSUE TITLE
Launch.java:62-64: Add exception handling to return...

### ISSUE THREAD

**Issue by karato** [Nerodesk/nerodesk#68]
Puzzle `14-7a1bdbb3` in `src/main/java/com/nerodesk/Launch.java:62-64` has to be resolved: Add exception handling to return 404 when resource not available. Now we show not nice stacktrace. Don't forget about unit tests.... The puzzle was created by longtimeago on 7-Mar-2015.

If you have any technical questions, don't ask me, submit new tickets instead

**Comment by karato** [comment]
@longtimeago do this task

**Comment by yegor256** [comment]
@longtimeago please wait for https://github.com/teamed/nerodesk/pull/77

**Comment by longtimeago** [comment]
@karato waiting for #77

**Comment by karato** [comment]
> @karato waiting for #77

@longtimeago yes, right, let's wait for #77

**Comment by karato** [comment]
Issue #77 is already closed (we were waiting for it) Maybe we can continue..

**Comment by longtimeago** [comment]
@yegor256 could you please clarify if this issue is still valid? Maybe some updates?

**Comment by yegor256** [comment]
@longtimeago I think it's a valid problem. try to open this page for example: http://beta.nerodesk.com/oops We should use `TsFallback` to prevent this.  see how it's done in Rultor.

**Comment by longtimeago** [comment]
@yegor256 What is in the scope? What is desired output of the task?
Major questions:
1) What exceptional situations should be handled by `TsFallback`?
2) Are we OK with one global handler?
3) What should client see if something bad happened?

I tried to encase only storage operations with `TsFallback` but looks like it works not stably.

**Comment by yegor256** [comment]
@longtimeago I think that a single fallback will be fine, just like it's done in rultor. I didn't get the "not stably" message. What is not stable and why?

**Comment by longtimeago** [comment]
@yegor256 Here is the code 

``` java
    public static Takes make(final Base base) throws IOException {
        final Takes fork = new TsFork(
            new FkParams(
                PsByFlag.class.getSimpleName(),
                Pattern.compile(".+"),
                new TkRedirect()
            ),
            new FkRegex(
                "/xsl/.*",
                new TsWithType(new TsClasspath(), "text/xsl")
            ),
            new FkRegex(
                "/css/.*",
                new TsWithType(
                    new TsFork(
                        new FkHitRefresh(
                            new File("./src/main/scss"),
                            new Runnable() {
                                @Override
                                public void run() {
                                    new VerboseProcess(
                                        new ProcessBuilder(
                                            "mvn",
                                            "sass:update-stylesheets",
                                            "minify:minify"
                                        )
                                    ).stdout();
                                }
                            },
                            new TsFiles("./target/classes")
                        ),
                        new FkFixed(new TsClasspath())
                    ),
                    "text/css"
                )
            ),
            new FkRegex("/robots.txt", ""),
            new FkRegex(
                "/",
                new TsFork(
                    new FkAuthenticated(
                        new Target<Request>() {
                            @Override
                            public Take route(final Request req)
                                throws IOException {
                                return new TkDocs(
                                    base.user(
                                        new RqAuth(req).identity().urn()
                                    ).docs(),
                                    req
                                );
                            }
                        }
                    ),
                    new FkAnonymous(
                        new Target<Request>() {
                            @Override
                            public Take route(final Request req) {
                                return new TkIndex(req);
                            }
                        }
                    )
                )
            ),
            new FkFixed(
                new TsFallback(
                    new TsFork(
                        new FkRegex(
                            "/r",
                            new Takes() {
                                @Override
                                public Take route(final Request req) throws IOException {
                                    final String file =
                                        new RqHref(req).href().param("f").iterator().next();
                                    return new TkRead(
                                        base.user(
                                            new RqAuth(req).identity().urn()
                                        ).docs().doc(file)
                                    );
                                }
                            }
                        ),
                        new FkRegex(
                            "/d",
                            new Takes() {
                                @Override
                                public Take route(final Request req) throws IOException {
                                    final String file =
                                        new RqHref(req).href().param("x").iterator().next();
                                    return new TkDelete(
                                        base.user(
                                            new RqAuth(req).identity().urn()
                                        ).docs().doc(file)
                                    );
                                }
                            }
                        ),
                        new FkRegex(
                            "/w",
                            new Takes() {
                                @Override
                                public Take route(final Request req) throws IOException {
                                    return new TkWrite(
                                        base.user(
                                            new RqAuth(req).identity().urn()
                                        ).docs(),
                                        req
                                    );
                                }
                            }
                        )
                    ),
                    new TkHTML("oops, something went wrong!")
                )
            )
        );
        return new TsFlash(App.auth(fork));
    }
```

Sorry for the long piece of code. What I've done - put all document's operations inside

```
            new FkFixed(
                new TsFallback(
                    new TsFork(
```

What is unstable? I see expected `oops, something went wrong!` only after very first call after app restart. On next calls I see just nothing. Empty body.

**Comment by yegor256** [comment]
@longtimeago brrr... what FkFixed is doing here? You should just wrap the entire `Takes` into `TsFallback`, just like Takes README explains. Do you understand how it works? It you don't, submit a bug to Takes github and I'll improve the documentation

**Comment by longtimeago** [comment]
PR is available https://github.com/teamed/nerodesk/pull/99

**Comment by longtimeago** [comment]
@karato we are done, please close the issue

**Comment by karato** [comment]
@longtimeago 7 days already, in 48 hours I will re-assign; as a reminder, don't forget to remove the `@todo` puzzle from the source code; -30 added to your rating, current score is: [+3410](http://www.netbout.com/b/35024?open=rating)

**Comment by longtimeago** [comment]
@karato we are done! please, close the issue!

**Comment by longtimeago** [comment]
@karato we are waiting for https://github.com/teamed/nerodesk/pull/99

**Comment by karato** [comment]
once `68-7fedbe9f` puzzle is resolved (later, in another ticket), this ticket will be fully complete

**Comment by longtimeago** [comment]
@karato please, close the issue. It's done

**Comment by karato** [comment]
Thanks all, I am closing

**Comment by karato** [comment]
@longtimeago thanks, paid, **30 mins** to your account, payment ID is `54203882`, task took 196 hours and 55 mins to complete

+30 to your rating, your total score is [+3610](http://www.netbout.com/b/35024?open=rating)
